src/node_http2.cc

```diff
@@ -1116,6 +1118,61 @@ void Http2Session::Unconsume() {
 }
 
 
+Headers::Headers(Isolate* isolate,
+                 Local<Context> context,
+                 Local<Array> headers) {
+  CHECK_EQ(headers->Length(), 2);
+  Local<Value> header_string = headers->Get(context, 0).ToLocalChecked();
+  Local<Value> header_count = headers->Get(context, 1).ToLocalChecked();
+  CHECK(header_string->IsString());
+  CHECK(header_count->IsUint32());
+  count_ = header_count.As<Uint32>()->Value();
+  size_t header_string_len = header_string.As<String>()->Length();
+
+  if (count_ == 0) {
+    CHECK_EQ(header_string_len, 0);
+    return;
+  }
+
+  // Allocate a single buffer with count_ nghttp2_nv structs, followed
+  // by the raw header data as passed from JS.
+  buf_.AllocateSufficientStorage(count_ * sizeof(nghttp2_nv) +
+                                 header_string_len);
+  char* header_contents = *buf_ + (count_ * sizeof(nghttp2_nv));
+  nghttp2_nv* const nva = reinterpret_cast<nghttp2_nv*>(*buf_);
+
+  CHECK_EQ(StringBytes::Write(isolate,
+                              header_contents, header_string_len,
+                              header_string, ASCII), header_string_len);
+
+  size_t n = 0;
+  char* p;
+  for (p = header_contents; p < header_contents + header_string_len; n++) {
+    if (n >= count_) {
+      // This can happen if a passed header contained a null byte. In that
+      // case, just provide nghttp2 with an invalid header to make it reject
+      // the headers list.
```

Comment:
@jasnell Yes, we could do that, but is that really a case that we want to optimize for? Plus, it would create inconsistency with how we handle other invalid headers.